Support multi source source provenance info

joshua-zivkovic opened the pull request:
Fixes https://github.com/apache/buildstream/issues/2073

This PR provides a way for multi-source plugins to provide source provenance info in their own way instead of only the top level source source info being used.

This is done by exposing the optional `source_provenance_attrs` argument to `create_source_info()` which when used overrides the usage of the default provenance for a source. Any plugin that needs to parse provenance in it's own way can then supply it's results using this. This has been trialled with the `cargo2` source plugin in https://gitlab.com/BuildStream/buildstream-plugins-community/-/merge_requests/408.

juergbi commented on src/buildstream/types.py:
```diff
@@ -164,6 +164,65 @@ class OverlapAction(FastEnum):
+            provenance["homepage"] = self.homepage
+
+        if self.issue_tracker is not None:
+            provenance["issue-tracker"] = self.homepage
```
```suggestion
            provenance["issue-tracker"] = self.issue_tracker
```

juergbi commented:
Do plugins need access to the actual `SourceProvenance` object or might it be sufficient to make the new `create_source_info()` parameter a `MappingNode` and let `create_source_info()` create a `_SourceProvenance` internally? This would expose less in the public API, but I'm not sure whether that would suffice or not.

joshua-zivkovic commented:
Yup, taking in to account https://github.com/apache/buildstream/pull/2095#issuecomment-3596947680, there's no need to expose SourceProvenance to achieve this.

juergbi committed:
source.py: Allow source provenance info to be overridden

This allows multi-source source plugins to provide this information per
source rather than as a singular top level.

This is done by adding a `provenance_node` parameter to
`create_source_info()` that when specified overrides the use of the
source's top level source provenance info.

Co-authored-by: Joshua Zivkovic <joshua.zivkovic@codethink.co.uk>

juergbi commented:
I've updated this branch as per my previous comment, but this still requires a test.

juergbi committed:
tests/frontend/show.py: Add source info test for multi-source plugin

joshua-zivkovic committed:
Add source provenance attribute tests

abderrahim commented on src/buildstream/source.py:
```diff
@@ -1367,6 +1367,7 @@ def create_source_info(
         version: str,
         *,
         version_guess: Optional[str] = None,
+        provenance_node: Optional[MappingNode] = None,
```
nit: I would add this after extra_data.

abderrahim commented on src/buildstream/source.py:
```diff
@@ -1386,15 +1387,23 @@ def create_source_info(
                          choice depicting the type of version.
            version: A string which represents a unique version of this source input
            version_guess: An optional string representing the guessed human readable version
+           provenance_node: The optional YAML node with source provenance attributes,
```
It's probably better to refer to the Node class rather than saying a "YAML node"

abderrahim commented on src/buildstream/source.py:
```diff
@@ -1386,15 +1387,23 @@ def create_source_info(
            version: A string which represents a unique version of this source input
            version_guess: An optional string representing the guessed human readable version
+           provenance_node: The optional YAML node with source provenance attributes,
+                            defaults to the provenance specified at the top level of the source.
```
I'm not sure this silent fallback semantic is good. IIUC, if the plugin provides a provenance for every source info, this will be silently ignored. If the plugin provides multiple source infos, and only some of them set provenance info (e.g. your implementation in https://gitlab.com/BuildStream/buildstream-plugins-community/-/merge_requests/408 if the user doesn't manually modify the ref to include provenance), then source infos that don't set it get the fallback of a global source provenance which is the same for everything.

I'm not sure what would be the best option here, but I feel the current implementation might not be a good idea.

joshua-zivkovic replied:
Ah, that is something that has been completely overlooked yes. I think a "multi-source" plugin should simply have a flag that it sets when used to tell the core to not use top-level source provenance data and actually disallow it (preventing the top level being silently ignored)

joshua-zivkovic replied:
I believe this has been resolved with the `BST_CUSTOM_SOURCE_PROVENANCE` flag being introduced to block usage of top level provenance definitions.

abderrahim commented on src/buildstream/source.py:
```diff
@@ -759,6 +759,14 @@ class Source(Plugin):
     # The defaults from the project
     __defaults: Optional[Dict[str, Any]] = None
 
+    BST_MULTI_SOURCE_PROVENANCE = False
```
I would call this CUSTOM or something. While the main use is for sources that have multiple source fetchers that might want to have different source provenance, the *effect* of setting this flag is to allow the source to set its own source provenance programatically and prevent the buildstream core from parsing / using the provenance information from the top-level source configuration.

abderrahim commented on src/buildstream/source.py:
```diff
@@ -1387,14 +1396,32 @@ def create_source_info(
            version: A string which represents a unique version of this source input
            version_guess: An optional string representing the guessed human readable version
            extra_data: Additional plugin defined key/values
+           provenance_node: The optional :class:`Node <buildstream.node.Node>` with source provenance attributes,
```
nit: The -> An

abderrahim commented on src/buildstream/source.py:
```diff
@@ -1387,14 +1396,32 @@ def create_source_info(
+                # This warning is printed for each occurrence of missing provenance
+                # for each sub-source, maybe it should be done just the once per
+                # multi-source?
+                self.warn("Multi source plugin: Not using top level source provenance")
```
As a follow-up to my comment above, this shouldn't call them "Multi source plugin".

Also I'm not sure if this should be a warning or an error. I feel we should be strict with this, especially if this check runs early on in the pipeline (i.e. it's not only run if are looking for source info) 

And please use the provenance (as in the `ProvenanceInformation`, not the `SourceProvenance`) in the error/warning message.

juergbi replied:
I think we should indeed make this an error and move it to the `Source` constructor. I don't see a backward compatibility concern as there are no `BST_CUSTOM_SOURCE_PROVENANCE` source plugins yet. If needed, we could potentially be less strict in the future, while the other way round would be impossible.

abderrahim commented on tests/frontend/show.py:
```diff
@@ -814,3 +814,33 @@ def test_source_info_workspace(cli, datafiles, tmpdir):
+    assert source_info.get_str("version-type") == "pony-age"
+    assert source_info.get_str("version") == "1234567"
+    assert source_info.get_str("version-guess", None) == "12"
+    assert source_info.get_str("homepage") == "https://flying-ponies.com/index.html"
```
Is this test up to date? Shouldn't the homepage here be unset?

(also should we test for the warning/error condition?)

joshua-zivkovic replied:
Yes, I'm currently updating the tests to match the new behaviour

juergbi replied:
The incorrect homepage check has been removed. I'd suggest adding
```
assert "homepage" not in source_info
```
to check that the homepage is actually unset.

joshua-zivkovic replied:
Added this in too, everything is happy.

juergbi commented on tests/frontend/source-info/elements/multisource.bst:
```diff
@@ -0,0 +1,7 @@
+- kind: multisource
+  ref:
+    - provenance:
+      homepage: https://flying-ponies.com/index.html
```
I don't think this is used, is it? An additional test with a test plugin that uses provenance from `ref` could be useful, but specifying it here without being used is confusing. I consider such an additional test a blocker as nice-to-have but not a blocker. If you add such a test, please keep the current test using `Node.from_dict` as we want to make sure that a plugin can also programmatically create provenance information from non-ref metadata.

joshua-zivkovic replied:
I'm not sure I entirely see the benefit of having a specific test for provenance being provided from `ref`, actually. It isn't something that is explicitly supported.

juergbi replied:
I'm fine with skipping that.

juergbi commented on tests/sources/source_provenance_attributes/project.conf:
```diff
@@ -0,0 +1,11 @@
+element-path: elements
+
+# Don't allow any provenance
+source-provenance-fields: {}
```
I think that test slipped in from the other PR.

joshua-zivkovic replied:
Yup, I'm about to push a fix

juergbi commented on src/buildstream/source.py:
```diff
@@ -828,6 +836,12 @@ def __init__(
 
+        if self.__provenance is not None and self.BST_CUSTOM_SOURCE_PROVENANCE:
+            raise SourceError(
+                f"{self._get_provenance()} Custom source plugin: Refusing to use top level source provenance",
```
"Custom source plugin" is not clear to me. Did you mean "Custom source provenance" or maybe "Custom source provenance plugin"?

joshua-zivkovic replied:
Ah, that is definitely not what I meant to put :D
